### PR TITLE
Implement /today endpoint

### PR DESCRIPTION
**Pull request by stishenok**
- [x] Implement /today endpoint
- [x] Implement /week endpoint
- [x] Implement /date endpoint
- [x] Implement /sun endpoint
- [x] Implement client/client.go app

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+	logger = log.WithFields(log.Fields{
+		"logger": "main",
+		"serviceContext": map[string]string{
+			"service": serviceName,
+			"version": serviceVersion,
+		},
+	})
+
+	mw = integration.Init(logger)
```
Comment (xedinaska) [inline client/client.go]: use integration full name please (`metaweather`)

File: client/client.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	fmt.Println(fmt.Sprintf("Today weather:\n%v\n", formatResponse(response)))
+}
+
+func getWeatherDate() {
+	response, err := mw.GetWeatherForDate(ctx, &api.DateWeatherRequest{
+		Latitude:  53.9045,
+		Longitude: 27.5615,
+		Date:      "2019-04-12",
```
Comment (xedinaska) [inline client/client.go]: after some week it won't work ('cause date will be too old)
I think you need to do smth like this: 

`const dateFormat = "2006-01-02"`
... 
`Date: time.Now().Format(dateFormat)`

File: client/client.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	fmt.Println(fmt.Sprintf("Week weather:\n%v\n", formatResponse(response)))
+}
+
+func getWeatherSun() {
+	response, err := mw.GetSunInfo(ctx, &api.SunInfoRequest{
+		Latitude:  53.9045,
+		Longitude: 27.5615,
+		Date:      "2019-04-12",
```
Comment (xedinaska) [inline client/client.go]: same here

File: client/client.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	})
+
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	fmt.Println(fmt.Sprintf("Sun weather for today:\n%v\n", formatResponse(response)))
```
Comment (xedinaska) [inline client/client.go]: Sun weather for today => Sun info for today

File: config/config.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package config
+
+const BaseURL = "https://www.metaweather.com/api"
```
Comment (xedinaska) [inline config/config.go]: you need to define env variable names here, not values. For example: 

`const BaseURL = 'API_URL'`

And then provide env varialbles with corresponding names: like `EXPORT API_URL=https://www.metaweather.com/api` 

in that case you'll be able switch your api url to staging / dev / prod / etc without any changes in your code

File: integration/common.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+	var weather weatherJSON
+	requestURL := fmt.Sprintf("%s%d/%d/%d/%d/", config.WeatherURL, locationID, date.Year(), date.Month(), date.Day())
+
+	if _, err := m.RequestClient.Get(ctx, requestURL, nil, &weather.ConsolidatedWeather); err != nil {
+		return nil, err
+	}
+
+	if len(weather.ConsolidatedWeather) == 0 {
+		return nil, errors.New("got zero size array in weather JSON")
```
Comment (xedinaska) [inline integration/common.go]: *customers-specific comment. 
They have a bit tricky errors system. Your error message should be smth like: 
```
err := errors.New("got zero size array in weather JSON")
return nil, errors.Wrap(err, "failed to receive weather by date")
```

in that case you: 
- provide info which call is failed 
- provide info why it was failed

File: integration/sun.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
+package integration
+
+import (
+	"context"
+	"fmt"
+	"github.com/stishenok/int-metaweather/config"
```
Comment (xedinaska) [inline integration/sun.go]: please use `goimports` before commit

File: integration/sun.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+		return nil, errors.Wrap(err, "failed to get location id")
+	}
+
+	// Get weather using location ID.
+	var weekWeather weatherJSON
+	if _, err := m.RequestClient.Get(ctx, fmt.Sprintf("%s%d/", config.WeatherURL, locationID), nil, &weekWeather); err != nil {
+		return nil, errors.Wrap(err, "failed to get weather using location_id")
+	}
+	if len(weekWeather.ConsolidatedWeather) == 0 {
+		return nil, errors.New("got zero size array in weather JSON")
```
Comment (xedinaska) [inline integration/sun.go]: same here

File: integration/today.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/stishenok/int-metaweather/config"
+	"github.com/xedinaska/int-weather-sdk/api"
+)
+
+type weatherJSON struct {
+	ConsolidatedWeather []struct {
```
Comment (xedinaska) [inline integration/today.go]: struct-in-struct isn't a good practice

File: main.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+	})
+)
+
+func main() {
+	conf, err := config.Load()
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+
+	metaweatherIntegration := integration.Init(logger)
```
Comment (xedinaska) [inline main.go]: `Integration` suffix isn't required here

File: client/client.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+			"service": serviceName,
+			"version": serviceVersion,
+		},
+	})
+
+	metaweather = integration.Init(logger)
+
+	var wg sync.WaitGroup
+	wg.Add(4)
+	go getWeatherToday(&wg)
```
Comment (xedinaska) [inline client/client.go]: you don't need to run them in routines. Please remove `go` part

File: config/config.go [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package config
+
+// Initialize this variable
+var API_URL string
+
+const (
+	// URLs
```
Comment (xedinaska) [inline config/config.go]: these comments haven't any sense, you can remove them

File: config/config.go [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+	BaseURL     = "API_URL"
+	LocationURL = "location/search/"
```
Comment (xedinaska) [inline config/config.go]: move these consts to `integration/metaweather.go`

File: integration/sun.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+func (m *Metaweather) GetSunInfo(ctx context.Context, req *api.SunInfoRequest) (*api.SunInfoResponse, error) {
+	// Get location ID using latitude and longitude coordinates.
+	locationID, err := m.getLocationID(ctx, req.Latitude, req.Longitude)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get location id")
+	}
+
+	// Get weather using location ID.
+	var weekWeather weatherJSON
+	if _, err := m.RequestClient.Get(ctx, fmt.Sprintf("%s%d/", config.WeatherURL, locationID), nil, &weekWeather); err != nil {
```
Comment (xedinaska) [inline integration/sun.go]: what if RequestClient won't return any err but will return 500 code in response.StatusCode?

File: integration/sun.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+		return nil, errors.Wrap(err, "failed to get weather using location_id")
+	}
+	if len(weekWeather.ConsolidatedWeather) == 0 {
+		err := errors.New("got zero size array in weather JSON")
+		return nil, errors.Wrap(err, "failed to receive sun info by date")
+	}
+
+	// check date
+	reqDate := req.Date
```
Comment (xedinaska) [inline integration/sun.go]: wat?

File: integration/today.go [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+type consolidatedWeather struct {
+	WeatherStateName string  `json:"weather_state_name"`
+	MinTemp          float64 `json:"min_temp"`
+	MaxTemp          float64 `json:"max_temp"`
+	WindSpeed        float64 `json:"wind_speed"`
+	Humidity         int     `json:"humidity"`
+	Visibility       float64 `json:"visibility"`
+	ApplicableDate   string  `json:"applicable_date"`
+}
+type weatherJSON struct {
```
Comment (xedinaska) [inline integration/today.go]: add empty line between struct definitions

File: integration/week.go [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+	// Get location ID using latitude and longitude coordinates.
+	locationID, err := m.getLocationID(ctx, req.Latitude, req.Longitude)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get location id")
+	}
+
+	var weekMetaWeather weatherJSON
+	// Get weather for 6 days using 1 request
+	if _, err := m.RequestClient.Get(ctx, fmt.Sprintf("%s%d/", config.WeatherURL, locationID), nil, &weekMetaWeather); err != nil {
```
Comment (xedinaska) [inline integration/week.go]: same here

File: integration/week.go [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+		return nil, errors.Wrap(err, "incorrect date")
+	}
+	lastDay = lastDay.AddDate(0, 0, 1)
+
+	weather, err := m.getWeatherByDate(ctx, locationID, lastDay)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get weather by date")
+	}
+
+	// 6+1 day
```
Comment (xedinaska) [inline integration/week.go]: make comment more descriptive

File: integration/sun.go (hunk 14 above)
Comment (stishenok) [reply integration/sun.go]: FATA[0000] failed to get location id: failed to unmarshal response body: unexpected end of JSON input  logger=main serviceContext="map[service:int-metaweather version:0.0.1]"

File: config/config.go [hunk 19]
```diff
@@ -0,0 +1,6 @@
+package config
+
+// Initialize this variable
+var API_URL string
```
Comment (xedinaska) [inline config/config.go]: https://github.com/apaliavy/golangcodestyle/#mixed-caps

File: client/client.go [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+)
+
+var ctx context.Context
+var logger *log.Entry
+var metaweather *integration.Metaweather
+
+func main() {
+	serviceName, serviceVersion := "int-metaweather", "0.0.1"
+
+	config.API_URL = os.Getenv(config.BaseURL)
```
Comment (xedinaska) [inline client/client.go]: you don't need to use global variable here (and in general - please avoid to use global variables: https://github.com/apaliavy/golangcodestyle/#avoid-global-variables) 
Use context to share this value

File: client/client.go [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+	if config.API_URL == "" {
```
Comment (xedinaska) [inline client/client.go]: => this condition should be also changed

File: integration/sun.go [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+func (m *Metaweather) GetSunInfo(ctx context.Context, req *api.SunInfoRequest) (*api.SunInfoResponse, error) {
+	// Get location ID using latitude and longitude coordinates.
+	locationID, err := m.getLocationID(ctx, req.Latitude, req.Longitude)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get location id")
+	}
+
+	// Get weather using location ID.
+	var weekWeather weatherJSON
+	if _, err := m.RequestClient.Get(ctx, fmt.Sprintf("%s%d/", WeatherURL, locationID), nil, &weekWeather); err != nil {
```
Comment (xedinaska) [inline integration/sun.go]: what if API returns 500 status code ?

File: config/config.go (hunk 19 above)
Comment (stishenok) [reply config/config.go]: Thank you, I remember it, but https://github.com/golang/go/wiki/CodeReviewComments#initialisms said that I can use only URL/url, API/api. APIURL/URLAPI not readable.
Anyway, delete the global variable. Thank you for your comment)

File: integration/sun.go (hunk 22 above)
Comment (stishenok) [reply integration/sun.go]: We will get error ("got 500 Internal Server Error").
https://github.com/stishenok/int-metaweather/blob/529cd5a147c360387fe6f443b134f9ede32ae9b1/request/request.go#L129

File: client/client.go (hunk 20 above)
Comment (stishenok) [reply client/client.go]: Added new string field in Metaweather struct:
https://github.com/stishenok/int-metaweather/pull/1/commits/ab58694ed873fe090c5481610620acae329ad41a#diff-59369197d95bddf910df1a6048a2b83c
When we receive request/call handler method, we are append url to ctx:
https://github.com/stishenok/int-metaweather/blob/ab58694ed873fe090c5481610620acae329ad41a/integration/sun.go#L14-L16
And read it in request.go:
https://github.com/stishenok/int-metaweather/blob/ab58694ed873fe090c5481610620acae329ad41a/request/request.go#L74

File: client/client.go (hunk 21 above)
Comment (stishenok) [reply client/client.go]: same here

File: config/config.go (hunk 19 above)
Comment (stishenok) [reply config/config.go]: Removed
https://github.com/stishenok/int-metaweather/blob/ab58694ed873fe090c5481610620acae329ad41a/config/config.go#L1-L3

File: integration/sun.go (hunk 22 above)
Comment (xedinaska) [reply integration/sun.go]: ok.. if API will return 301, 401, 404, 422 ... etc - any value <> 200? 

common approach that we use is: 
```
response, err := m.RequestClient.Get(..) 
if err != nil {
... //handle error 
}

if response.StatusCode != http.StatusOK {
... //return err 
}

//happy-path here 
```

File: integration/sun.go [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+		return nil, errors.Wrap(err, "failed to get weather using location_id")
+	}
+	if len(weekWeather.ConsolidatedWeather) == 0 {
+		err := errors.New("got zero size array in weather JSON")
+		return nil, errors.Wrap(err, "failed to receive sun info by date")
+	}
+
+	// Check date (we can accept only current date)
+	if req.Date != weekWeather.ConsolidatedWeather[0].ApplicableDate {
+		return nil, fmt.Errorf("we can provide sun/set information only for today (%s)", weekWeather.ConsolidatedWeather[0].ApplicableDate)
```
Comment (xedinaska) [inline integration/sun.go]: better solution: receive date from request, check: is it today's date or nope, if nope - return error without sending request to external API. It'll work much more faster

File: integration/sun.go [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+	// Append url value to context.
+	ctx = context.WithValue(ctx, config.BaseURL, m.URL)
+	// Get location ID using latitude and longitude coordinates.
+	locationID, err := m.getLocationID(ctx, req.Latitude, req.Longitude)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get location id")
+	}
+
+	// Get weather using location ID.
+	var weekWeather weatherJSON
```
Comment (xedinaska) [inline integration/sun.go]: better: weekWeather := &weatherJSON{} or weekWeather := new(weatherJSON)

File: integration/today.go [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+
+	// Get location ID using latitude and longitude coordinates.
+	locationID, err := m.getLocationID(ctx, req.Latitude, req.Longitude)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get location id")
+	}
+
+	// Get weather using location ID.
+	var weekWeather weatherJSON
+	if _, err := m.RequestClient.Get(ctx, fmt.Sprintf("%s%d/", WeatherURL, locationID), nil, &weekWeather); err != nil {
```
Comment (xedinaska) [inline integration/today.go]: same here: need to add response status code check

File: integration/week.go [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+
+	// Get location ID using latitude and longitude coordinates.
+	locationID, err := m.getLocationID(ctx, req.Latitude, req.Longitude)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get location id")
+	}
+
+	var weekMetaWeather weatherJSON
+	// Get weather for 6 days using 1 request
+	if _, err := m.RequestClient.Get(ctx, fmt.Sprintf("%s%d/", WeatherURL, locationID), nil, &weekMetaWeather); err != nil {
```
Comment (xedinaska) [inline integration/week.go]: and here

File: integration/sun.go (hunk 14 above)
Comment (xedinaska) [reply integration/sun.go]: see my comments below

File: integration/sun.go (hunk 22 above)
Comment (stishenok) [reply integration/sun.go]: Now got it
I had a question why we check only 500th error)

File: integration/sun.go (hunk 23 above)
Comment (stishenok) [reply integration/sun.go]: resolved in skype

**Pull request merged**